pkg/fs/probe.go

```diff
@@ -0,0 +1,154 @@
+/*
+ * This file is part of MinIO Direct CSI
+ * Copyright (c) 2021 MinIO, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package fs
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	fserrors "github.com/minio/direct-csi/pkg/fs/errors"
+	"github.com/minio/direct-csi/pkg/fs/ext4"
+	"github.com/minio/direct-csi/pkg/fs/fat32"
+	"github.com/minio/direct-csi/pkg/fs/swap"
+	"github.com/minio/direct-csi/pkg/fs/xfs"
+)
+
+// FS denotes filesystem interface.
+type FS interface {
+	ID() string
+	Type() string
+	TotalCapacity() uint64
+	FreeCapacity() uint64
+}
+
+func probe(device string) (FS, error) {
+	devFile, err := os.OpenFile(device, os.O_RDONLY, os.ModeDevice)
```

Comment:
Suggestion - earlier we had the following style

```
filesystems := []Filesystem{
		xfs.NewXFS(),
		ext4.NewEXT4(),
		fat32.NewFAT32(),
		swap.NewSwap(),
		// Add new filesystems here
	}

	for _, fs := range filesystems {
		found, err := fs.ProbeFS(device, 0)
		if err != nil {
			return nil, err
		}
		if found {
			return fs, nil
		}
	}
```

Can we keep this style if possible?

Reply:
No. Each filesystem probing accepts different arguments. In current code, `fat32` requires `io.ReadSeeker` whereas others require `io.Reader`. Current is working because each filesystem opens device independently. This is what fixed here.